src/understand-tidb/transaction.md

```diff
@@ -1 +1,111 @@
+	// Valid returns if the transaction is valid.
+	// A transaction become invalid after commit or rollback.
+	Valid() bool
+	// GetMemBuffer return the MemBuffer binding to this transaction.
+	GetMemBuffer() MemBuffer
+	// GetSnapshot returns the Snapshot binding to this transaction.
+	GetSnapshot() Snapshot
+  ...
+}
+```
+
+These are common interfaces the transaction will provide, for example `Commit` will be used to commit the current ongoing transaction. The transaction is considered ongoing before the `commit` operations is triggered. The two phase commit processing will be used to commit a transaction and it will finally become committed or aborted. 
+
+[LazyTxn](https://github.com/pingcap/tidb/blob/master/session/txn.go#L50) is a wrapper of the transaction implementations, when the SQL statements are executed using a stand-alone session context, `LazyTxn` will be used to do things like：
+- Return the memory buffer for write.
+- Set specific operations or flags for the current transaction.
+- Return the snapshot of this transaction.
+- Commit the current transaction.
+- Lock specific keys.
+- etc...
+
+
+# The Statement Execution
+
+Usually the first thing that will be done executing a statment is to `activiate` the related transaction. By default `TiDB` proviedes the snapshot isolation level so in each new transction, a new global strong snapshot will be fetched first before executing statements. In `tidb` the snapshot is repsented by a global tso which is fetched from the `pd` server, and it acts as the unique idetifier for this transaction. After this operation a transaction is regarded as `activated`.
+
+For the read SQL statements, the [snapshot](https://github.com/pingcap/tidb/blob/master/store/driver/txn/snapshot.go) will be used to provide a global strong consistent snapshot, all the reads will check data visibility using this snapshot. Most executors will set the timestamp doing the build, and the transaction could be activiated by the building process. Some commonly used [snapshot](https://github.com/pingcap/tidb/blob/master/store/driver/txn/snapshot.go#L40) API:
+```
+// BatchGet gets all the keys' value from kv-server and returns a map contains key/value pairs.
+// The map will not contain nonexistent keys.
+func (s *tikvSnapshot) BatchGet(ctx context.Context, keys []kv.Key) (map[string][]byte, error) {
+	data, err := s.KVSnapshot.BatchGet(ctx, toTiKVKeys(keys))
+	return data, extractKeyErr(err)
+}
+
+// Get gets the value for key k from snapshot.
+func (s *tikvSnapshot) Get(ctx context.Context, k kv.Key) ([]byte, error) {
+	data, err := s.KVSnapshot.Get(ctx, k)
+	return data, extractKeyErr(err)
+}
+```
+
+For the write SQL statements, they will write data into the transaction memory buffer temporarily until the `commit` operations is triggered. There are 3 main interfaces which will write query data into the memory buffer, they are the [tables](https://github.com/pingcap/tidb/blob/master/table/table.go#L166) API:
```

Comment:
> https://github.com/pingcap/tidb/blob/master/table/table.go#L166

I'd prefer a permalink since line number of mater is unstable.

Reply:
@cfzjywxk what do you think?

Reply:
I will check and change the links, thanks for the remind.